src/Containers/Containers.jl

```diff
@@ -0,0 +1,226 @@
+#  Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#############################################################################
+# JuMP
+# An algebraic modeling language for Julia
+# See http://github.com/JuliaOpt/JuMP.jl
+#############################################################################
+
+"""
+    Containers
+
+Module defining the containers `SparseAxisArray` that behaves as a regular
+`AbstractArray` but with custom indexes that are not necessarily integers.
+"""
+module Containers
+
+"""
+    struct SparseAxisArray{T,N,K<:NTuple{N, Any}} <: AbstractArray{T,N}
+        data::Dict{K,T}
+    end
+
+`N`-dimensional array with elements of type `T` where only a subset of the
+entries are defined. The entries with indices `idx = (i1, i2, ..., iN)` in
+`keys(data)` has value `data[idx]`. Note that as opposed to
+`SparseArrays.AbstractSparseArray`, the missing entries are not assumed to be
+`zero(T)`, they are simply not part of the array. This means that the result of
+`map(f, sa::SparseAxisArray)` or `f.(sa::SparseAxisArray)` has the same sparsity
+structure than `sa` even if `f(zero(T))` is not zero.
+"""
+struct SparseAxisArray{T,N,K<:NTuple{N, Any}} <: AbstractArray{T,N}
+    data::Dict{K,T}
+end
+
+Base.length(sa::SparseAxisArray) = length(sa.data)
+Base.IteratorSize(::Type{<:SparseAxisArray}) = Base.HasLength()
+# By default `IteratorSize` for `Generator{<:AbstractArray{T,N}}` is
+# `HasShape{N}`
+Base.IteratorSize(::Type{Base.Generator{<:SparseAxisArray}}) = Base.HasLength()
+# Needed in `collect_to_with_first!`
+Base.eachindex(g::Base.Generator{<:SparseAxisArray}) = eachindex(g.iter)
+Base.iterate(sa::SparseAxisArray, args...) = iterate(values(sa.data), args...)
+# Used by JuMP.getloopedcode
+Base.haskey(sa::SparseAxisArray, idx) = haskey(sa.data, idx)
+function Base.haskey(sa::SparseAxisArray{T,1,Tuple{I}}, idx::I) where {T, I}
+    return haskey(sa.data, (idx,))
+end
+
+# A `length` argument can be given because `IteratorSize` is `HasLength`
+function Base.similar(sa::SparseAxisArray{S,N,K}, ::Type{T},
+                      length::Integer=0) where {S, T, N, K}
+    d = Dict{K, T}()
+    if !iszero(length)
+        sizehint!(d, length)
+    end
+    return SparseAxisArray(d)
```

Comment:
How does the `N` parameter get inferred? Through `K`?

Reply:
Yes. I have added many `@inferred` in the tests to make sure it works